### DISCUSSION TITLE
How to get content from not main frame?

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3829] [Q&A]
I found solution here https://stackoverflow.com/questions/36105240/cefsharp-and-frames-only-retrieving-html-from-first-frame with clicking on frame content. Maybe there are any out of box solutions?

**Comment by amaitland** [comment]
You can get frames by their name or their ID.

http://cefsharp.github.io/api/94.4.x/html/Overload_CefSharp_IBrowser_GetFrame.htm

You can access the focused frame as well.

http://cefsharp.github.io/api/94.4.x/html/P_CefSharp_IBrowser_FocusedFrame.htm

**Reply by assopri** [reply]
Thank you - that is what I was searching for. However when I try to case my IWebBrowser object to IBrowser I get null:
`IBrowser browser = (myIWebBrowser as IBrowser);`

On the other hand, when I try to get IBrowser object by GetBrowser() method of my IWebBrowser, it does not identify frame on the page (getting null after getframe, despite I previously check frame existance with js):

`
if(!ElementExists("//iframe[@id='profitbase_front_widget']"))
            {
                throw new Exception("frame not exists on page");
            }

IFrame frame = myIWebBrowser.GetBrowser().GetFrame("profitbase_front_widget");//
`

What am I doing wrong?

**Reply by amaitland** [reply]
It's likely the name of the frame is different to the ID you are passing in. Try getting all the frame names and see if it matches what you are expecting.

**Reply by assopri** [reply]
So my progress is the following. On this webpage https://dom-kit.ru/#/profitbase/house/32806/bigGrid?filter=property.status:AVAILABLE I try to execute the following code. First I (successfully) check if iframe with the needed id exists, than I get all frame identifiers, then trying to iterate them and find needed frame. There are 2 of them, however I cannot find nor frame with this name "profitbase_front_widget" (that is still logical, because its not a name, but id), neither frame with content I see in my Chrome web browser:

![image](https://user-images.githubusercontent.com/73235464/135588726-d0af82f4-eace-4bd5-b411-3c9e1231a2f0.png)


`if(!cefBrowserWrapper.ElementExists("//iframe[@id='profitbase_front_widget']"))
            {
                throw new Exception("need wait");
            }
            IWebBrowser myIWebBrowser = cefBrowserWrapper.GetBrowser();

            var identifiers = myIWebBrowser.GetBrowser().GetFrameIdentifiers();

            foreach (var i in identifiers)
            {
                frame = myIWebBrowser.GetBrowser().GetFrame(i);
                if (frame.Name == "profitbase_front_widget")
                    return frame;
            }`

By the way, frame name from first identifier is empty, and from second - it is ""

Any ideas how can I find frame I need?

**Reply by assopri** [reply]
Update!!! The frame with the second identifier is what I need, however the content is encoded , so I cannot find data I can find in chrome web version:
![image](https://user-images.githubusercontent.com/73235464/135590073-04cc6f2c-f6c7-4401-89dd-65c6bd44fcad.png)

**Reply by amaitland** [reply]
> however the content is encoded

What do you mean by encoded? How are you trying to access the data exactly? If you evaluate JavaScript you should get content directly from the DOM.

**Reply by assopri** [reply]
> > however the content is encoded
> 
> What do you mean by encoded? How are you trying to access the data exactly? If you evaluate JavaScript you should get content directly from the DOM.

I researched more and found that I just do not wait for the entire frame content to be loaded. By the way, how can I wait for the particular dom element (with specific ID), which is included in frame, is loaded? (to continue working with webpage after it is loaded)

**Reply by assopri** [reply]
SOLVED!
